File: mas_garage/orchid_garage_dashboard/models/garage_dashboard.py
from odoo import api, models


class GarageDashboard(models.AbstractModel):
    _name = "garage.dashboard"

    @api.model
    def get_dashboard_data(self):
        SaleOrder = self.env["sale.order"]
        Estimation = self.env["garage.estimation"]
        Invoice = self.env["account.move"]

        # ── Job Orders ────────────────────────────────────────────────────────
        pending = SaleOrder.search([
            ("od_job_status", "in", ["draft", "in_progress"])
        ])
        closed = SaleOrder.search([
            ("od_job_status", "=", "closed")
        ])
        invoice_pending = SaleOrder.search([
            ("od_job_status", "=", "closed"),
            ("invoice_status", "=", "to invoice"),
        ])

        # ── Estimations ───────────────────────────────────────────────────────
        est_draft = Estimation.search([("state", "=", "draft")])
        est_confirmed = Estimation.search([("state", "=", "confirmed")])
        est_completed = Estimation.search([("state", "=", "completed")])

        # ── Invoices (customer invoices only) ─────────────────────────────────
        inv_posted = Invoice.search([
            ("move_type", "=", "out_invoice"),
            ("payment_state", "not in", ["paid", "in_payment"]),
            ("state", "=", "posted"),
        ])
        inv_paid = Invoice.search([
            ("move_type", "=", "out_invoice"),
            ("payment_state", "in", ["paid", "in_payment"]),
        ])
        inv_overdue = Invoice.search([
            ("move_type", "=", "out_invoice"),
            ("state", "=", "posted"),
            ("payment_state", "not in", ["paid", "in_payment"]),
            ("invoice_date_due", "<", fields_date()),
        ])

        return {
            # jobs
            "pending_job_count": len(pending),
            "closed_job_count": len(closed),
            "closed_amount_total": sum(closed.mapped("amount_total")),
            "invoice_pending_count": len(invoice_pending),
            "invoice_pending_amount_total": sum(invoice_pending.mapped("amount_to_invoice")),
            # estimations
            "est_draft_count": len(est_draft),
            "est_confirmed_count": len(est_confirmed),
            "est_completed_count": len(est_completed),
            "est_confirmed_amount": sum(est_confirmed.mapped("amount_total")),
            "est_completed_amount": sum(est_completed.mapped("amount_total")),
            # invoices
            "inv_posted_count": len(inv_posted),
            "inv_posted_amount": sum(inv_posted.mapped("amount_residual")),
            "inv_paid_count": len(inv_paid),
            "inv_paid_amount": sum(inv_paid.mapped("amount_total")),
            "inv_overdue_count": len(inv_overdue),
            "inv_overdue_amount": sum(inv_overdue.mapped("amount_residual")),
            # misc
            "currency_symbol": self.env.company.currency_id.symbol,
        }


def fields_date():
    from datetime import date
    return date.today().isoformat()

File: asp/orchid_garage/static/src/js/garage_dashboard.js
/** @odoo-module **/

import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { Component, useState, onWillStart } from "@odoo/owl";

// ── KPI Card ─────────────────────────────────────────────────────────────────
class KpiCard extends Component {
    static template = "orchid_garage_dashboard.KpiCard";
    static props = {
        label: String,
        value: [String, Number],
        sub: { type: String, optional: true },
        icon: String,
        color: String,
        onClick: { type: Function, optional: true },
        loading: { type: Boolean, optional: true },
    };
}

// ── Dashboard ─────────────────────────────────────────────────────────────────
class GarageDashboard extends Component {
    static template = "orchid_garage_dashboard.Dashboard";
    static components = { KpiCard };

    setup() {
        this.orm = useService("orm");
        this.action = useService("action");

        this.state = useState({
            loading: true,
            data: {
                pending_job_count: 0,
                closed_job_count: 0,
                closed_amount_total: 0,
                invoice_pending_count: 0,
                invoice_pending_amount_total: 0,
                est_draft_count: 0,
                est_confirmed_count: 0,
                est_completed_count: 0,
                est_confirmed_amount: 0,
                est_completed_amount: 0,
                inv_posted_count: 0,
                inv_posted_amount: 0,
                inv_paid_count: 0,
                inv_paid_amount: 0,
                inv_overdue_count: 0,
                inv_overdue_amount: 0,
            },
        });

        onWillStart(async () => { await this._loadData(); });
    }

    async _loadData() {
        this.state.loading = true;
        try {
            this.state.data = await this.orm.call("garage.dashboard", "get_dashboard_data", []);
        } catch (e) {
            console.error(e);
        } finally {
            this.state.loading = false;
        }
    }

    // ── Formatters ────────────────────────────────────────────────────────────
    fmt(value) {
        return (value !== undefined && value !== null) ? String(value) : "—";
    }

    fmtMoney(value) {
        const symbol = this.state.data.currency_symbol || "";
        return `${symbol} ${new Intl.NumberFormat().format(value || 0)}`;
    }

    // ── Job Order actions ─────────────────────────────────────────────────────
    async openPendingJobs() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Pending Jobs",
            res_model: "sale.order",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["od_job_status", "in", ["in_progress", "draft"]]],
        });
    }

    async openClosedJobs() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Closed Jobs",
            res_model: "sale.order",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["od_job_status", "=", "closed"]],
        });
    }

    async openInvoicePending() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Invoice Pending",
            res_model: "sale.order",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["od_job_status", "=", "closed"], ["invoice_status", "=", "to invoice"]],
        });
    }

    // ── Estimation actions ────────────────────────────────────────────────────
    async openEstDraft() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Draft Estimations",
            res_model: "garage.estimation",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "=", "draft"]],
        });
    }

    async openEstConfirmed() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Confirmed Estimations",
            res_model: "garage.estimation",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "=", "confirmed"]],
        });
    }

    async openEstCompleted() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Completed Estimations",
            res_model: "garage.estimation",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["state", "=", "completed"]],
        });
    }

    // ── Invoice actions ───────────────────────────────────────────────────────
    async openInvPosted() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Open Invoices",
            res_model: "account.move",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["move_type", "=", "out_invoice"], ["state", "=", "posted"],
                     ["payment_state", "not in", ["paid", "in_payment"]]],
        });
    }

    async openInvPaid() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Paid Invoices",
            res_model: "account.move",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [["move_type", "=", "out_invoice"], ["payment_state", "in", ["paid", "in_payment"]]],
        });
    }

    async openInvOverdue() {
        await this.action.doAction({
            type: "ir.actions.act_window",
            name: "Overdue Invoices",
            res_model: "account.move",
            view_mode: "list,form",
            views: [[false, "list"], [false, "form"]],
            domain: [
                ["move_type", "=", "out_invoice"],
                ["state", "=", "posted"],
                ["payment_state", "not in", ["paid", "in_payment"]],
                ["invoice_date_due", "<", new Date().toISOString().split("T")[0]],
            ],
        });
    }

    async refresh() { await this._loadData(); }
}

registry.category("actions").add("orchid_garage_dashboard", GarageDashboard);